File: apps/ios/SignalASI/VoiceTTSRequestRegistry.swift
import Foundation

struct VoiceTTSRequest {
  var utteranceId: String
  var traceId: String
  var onFinished: () -> Void
}

final class VoiceTTSRequestRegistry {
  private var active: VoiceTTSRequest?
  private let lock = NSLock()

  func begin(_ request: VoiceTTSRequest) {
    locked {
      active = request
    }
  }

  func finish(_ utteranceId: String?) -> VoiceTTSRequest? {
    locked {
      guard let current = active,
            matches(current, utteranceId: utteranceId) else {
        return nil
      }
      active = nil
      return current
    }
  }

  func discard(_ utteranceId: String?) -> Bool {
    locked {
      guard let current = active,
            matches(current, utteranceId: utteranceId) else {
        return false
      }
      active = nil
      return true
    }
  }

  func isActive(_ utteranceId: String?) -> Bool {
    locked {
      guard let current = active else { return false }
      return matches(current, utteranceId: utteranceId)
    }
  }

  func clear() {
    locked {
      active = nil
    }
  }

  private func matches(_ request: VoiceTTSRequest, utteranceId: String?) -> Bool {
    guard let utteranceId,
          !utteranceId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      return false
    }
    return request.utteranceId == utteranceId
  }

  private func locked<T>(_ body: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return body()
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/voice/VoiceTtsRequestRegistry.kt
package com.galaxyssi.chat.voice

internal data class VoiceTtsRequest(
    val utteranceId: String,
    val traceId: String,
    val onFinished: () -> Unit
)

internal class VoiceTtsRequestRegistry {
    private var active: VoiceTtsRequest? = null

    @Synchronized
    fun begin(request: VoiceTtsRequest) {
        active = request
    }

    @Synchronized
    fun finish(utteranceId: String?): VoiceTtsRequest? {
        val current = active ?: return null
        if (utteranceId.isNullOrBlank() || current.utteranceId != utteranceId) return null
        active = null
        return current
    }

    @Synchronized
    fun discard(utteranceId: String?): Boolean {
        val current = active ?: return false
        if (utteranceId.isNullOrBlank() || current.utteranceId != utteranceId) return false
        active = null
        return true
    }

    @Synchronized
    fun clear() {
        active = null
    }
}
